ls1intum/Artemis
# Do not specify exercise dates in exam exercises

### Checklist

### Motivation and Context
It is cumbersome and error prone to set all exercise dates (release, due, assessment due) for exams correctly.
Also students with individual working times cannot be reflected which is a problem.
It makes more sense to simply ignore these exercises dates and set them to null to prevent inconsistencies or other issues.
Exams should 

### Description
Instructors cannot set these dates any more when creating or editing exercises for exams.

* Do not allow users to specify exercise dates in exam exercises
* Always use the specified exam dates instead 
  * exam start date --> release date
  * exam end date (individual) --> due date
  * exam result publication date --> assessment due date
* Always set exercise dates in exam exercises to null
* Also allow to specify assessment due date in programming exercises

**Basically refactoring to go through the usage of all the dates in exercise in the current codebase and see if it needs to be adapted to support exam exercises (where all the dates are null)**

Open in the implementation:

- [x] Make sure that students can still submit all exercise types in their (personal) exam working time
- [x] Make sure the assessment works, when the exercise dates are null
- [x] Check that all tests are running properly (some might need to be adapted)

Known issues/open points to do in follow-up:
- allow to specify assessment due date in programming exercises
- fix tutor-exercise-dashboard leaderboard and side-panel
- make sure that tutors cannot assess submissions with no result (first assessment) after the result release date

### Steps for Testing
1. Create an exam with text, modeling, quiz and programming exercises (try out create, edit and import)
2. If possible, double check that release date, due date and assessment due date are set to null in the database
3. Participate as student in the exam and make sure you can submit until the end of the individual working time without any issues
4. Participate as tutor in the assessment of the exam. This should also work without any problems after the end date of the exam (or more precise the latest individual end date of all students). The Tutor Exam Dashboard will be visible after the latest individual end date.
5. After the result release date, a tutor should not be able to override his assessment (403 error should be thrown)

**Please also review the code properly**, there were some major refactoring for the `AssessmentResource` and `AssessmentService`
